src/main/java/org/minions/devfund/bruno/battleship/Ship.java

```diff
@@ -0,0 +1,264 @@
+package org.minions.devfund.bruno.battleship;
+
+import java.util.stream.IntStream;
+
+/**
+ * Ship will be an abstract class.
+ * We never really want to create a Ship. We always want to create a specific type of Ship.
+ * So this means the declaration will be abstract.
+ */
+public abstract class Ship {
+
+    private static final int SIZE = 20;
+    private int bowRow;
+    private int bowColumn;
+    private int length;
+    private boolean horizontal;
+    private boolean[] hit;
+
+    /**
+     * This is an abstract method and therefore has no body.
+     *
+     * @return String ship type name.
+     */
+    public abstract String getShipType();
+
+    /**
+     * Returns true if it is okay to put a ship of this length with its bow in this location,
+     * with the given orientation, and returns false otherwise.
+     * The ship must not overlap another ship, or touch another ship (vertically, horizontally).
+     *
+     * @param row        integer value.
+     * @param column     integer value
+     * @param horizontal boolean value..
+     * @param ocean      Ocean object.
+     * @return boolean value.
+     */
+    public boolean okToPlaceShipAt(final int row, final int column, final boolean horizontal, final Ocean ocean) {
+        int columnLength = column + getLength();
+        int rowLength = row + getLength();
+        return horizontal ? columnLength <= SIZE && !placeShipMatch(row, column, ocean, columnLength, row)
+                : rowLength <= SIZE && !placeShipMatch(row, column, ocean, column, rowLength);
+    }
+
+    /**
+     * The ship must not overlap another ship, or match with another ship in this column and row.
+     *
+     * @param row       integer value.
+     * @param column    integer value.
+     * @param ocean     Ocean object.
+     * @param colLength integer column length value.
+     * @param rowLength integer row length value.
+     * @return boolean value.
+     */
+    private boolean placeShipMatch(final int row, final int column, final Ocean ocean,
+                                   final int colLength, final int rowLength) {
+        return IntStream.rangeClosed(row - 1, rowLength + 1).anyMatch(i
+                -> IntStream.range(column - 1, colLength + 1).anyMatch(j -> handlingArrayIndex(ocean, i, j)));
+    }
+
+    /**
+     * It must not 'stick out' beyond the array handling the index out of bounds exception.
+     *
+     * @param ocean Ocean object.
+     * @param row   integer value.
+     * @param col   integer value.
+     * @return boolean value.
+     */
+    private boolean handlingArrayIndex(final Ocean ocean, final int row, final int col) {
+        boolean arrayIndex = false;
+        try {
+            arrayIndex = !ocean.getShipArray()[row][col].getShipType().equals("empty");
+        } catch (IndexOutOfBoundsException e) {
```

Comment:
Why are you using an IndexOutOfBoundsException ?
Could you handle it in previous steps?